packages/toolpad-app/src/toolpad/AppEditor/ConnectionEditor/index.tsx

```diff
@@ -49,18 +49,21 @@ function ConnectionEditorContent<P>({
   className,
   connectionNode,
 }: ConnectionEditorContentProps<P>) {
+  const { dom } = useDom();
   const domApi = useDomApi();
 
   const handleConnectionChange = React.useCallback(
     (connectionParams: P | null) => {
-      domApi.setNodeNamespacedProp(
+      const updatedDom = appDom.setNodeNamespacedProp(
+        dom,
         connectionNode,
         'attributes',
         'params',
         appDom.createSecret(connectionParams),
       );
+      domApi.update(updatedDom);
     },
-    [connectionNode, domApi],
+    [connectionNode, dom, domApi],
```

Comment:
🙂 I think this is why I suggested to use a function in `domApi.update`. This callback invalidates on every dom change, and it's also not 100% sure it will get the last `dom`. e.g. imagine a function.
```tsx
const foo = () => domApi.update(doSomethingWith(dom));
const bar = () => domApi.update(doSomethingElseWith(dom));

// somewhere else, foo called twice in a row
foo()
bar()
```
Those would get the same initial version of `dom`, while I would expect in most cases that in the call of `bar`, there should be an update on the result of `doSomethingWith(dom)` from call of `foo`

Reply:
Would it work to restore and use methods such as `domApi.setNodeNamespacedProp` (which I had deleted) in cases like this? 
And use `domApi.update` as I'm using it but only when a temporary DOM is needed? `domApi.update` replaces the whole DOM anyway so in those cases maybe it's fine?

Reply:
Still thinking of better alternatives, your proposal does seem more "atomic", it's just more difficult to use in `RenderOverlay`.

Reply:
I'll try supporting `domApi.update((draft) => ...` and use it as much as possible when I can, except in the few cases where it's hard to use, I think that should be possible.

Reply:
> Still thinking of better alternatives, your proposal does seem more "atomic", it's just more difficult to use in `RenderOverlay`.

We can support both methods, like 

```tsx
update: (draft: React.SetStateAction<AppDom>) => void
```

Reply:
Looks like I was able to use an updater function everywhere after all!
https://github.com/mui/mui-toolpad/pull/1374/commits/1f7779e2e9d097c6d5607e11915a59d180e39bc3
After the refactoring I had made I guess it wasn't so difficult.

Reply:
As this PR is using the updater function now I will merge this part already!